manager.c

```diff
@@ -297,6 +300,43 @@ static char * espace_newlines(const char * text)
 	return escaped;
 }
 
+#/* */
+static char * unescape_newlines(const char * text)
+{
+	char * unescaped;
+	int i, j;
+	for(j = i = 0; text[i]; ++i)
+	{
+		if(text[i] != '\\' || !(text[i + 1] == 'r' || text[i + 1] == 'n'))
+			j++;
+	}
+	unescaped = ast_malloc(j + 1);
+
+	if(unescaped)
+	{
+		for(j = i = 0; text[i]; ++i, ++j)
+		{
+			if(text[i] == '\\' && text[i + 1] == 'r')
+			{
+				unescaped[j] = '\r';
+				i++;
+			}
+			else if(text[i] == '\\' && text[i + 1] == 'n')
+			{
+				unescaped[j] = '\n';
+				i++;
+			}
+			else
+			{
+				unescaped[j] = text[i];
+			}
```

Comment:
This form of escaping, where the escape is not escapable, is not acceptable.

With this code, you cannot write: `"the line feed is \\n"` because it will turn into `"the line feed is\<LF>"`

Further, you'll want to keep this compatible with what Asterisk does. Asterisk indeed uses c-escaping for the AMI in various places, so backslash-escaping is good:
https://github.com/asterisk/asterisk/blob/9c2871edf4df413fdb43dbaee063677b970f21a2/main/presencestate.c#L531

Asterisk has c_escape/c_unescape functions:
https://github.com/asterisk/asterisk/blob/9c2871edf4df413fdb43dbaee063677b970f21a2/include/asterisk/strings.h#L319
I haven't checked whether they're exposed, but at least you'll want chan-dongle to behave the same.

Reply:
They are exposed and working as expected